requests-for-comments/2023-02-15-improving-our-client-libraries.md

```diff
@@ -0,0 +1,63 @@
+
+What's the decision: Should we try 2 months of working with contractors on improving the developer experience and getting to feature parity for the libraries. If this works, we will continue to work with the contractors. If not, we will re-evaluate.
+
+## Problems
+
+Developer experience problems:
+
+- Incomplete CI/CD
+- Lack of test coverage
+- Random legacy dependencies/development environments that don’t work
+- Patchy documentation
+- “Cross compatibility nuances (kotlin -> java, swift -> objc)”
+
+Custom problems:
+
+- Lack of 1st class support for React and [NextJS](https://github.com/PostHog/posthog-js/issues/492) (very common for high-potential and high-growth customers)
+- Missing feature parity for a number of libraries:
+    - [https://posthog.com/docs/integrate/libraries](https://posthog.com/docs/integrate/libraries)
+
+## Where we want to get to
+
+- Test coverage, CI/CD, and quality docs across all the repos
+- PostHog core team implements the features for the “Primary” libraries (the ones we use ourselves): Javascript, Python and Node.js
+- Contractors implement new features and the core team reviews the PRs for the “Secondary” libraries (all the other ones). Aiming to keep feature parity within a few weeks
+- Each team is responsible for reviewing their features.
+- TBD who reviews the core dependencies? Should we instead go with individuals that have good knowledge of those libraries
+- Luke to get the system setup for working with the contractors via a shared slack and github project. Each team as they ship features work directly with the contractors for feature implementation. Luke responsible for making sure the ongoing core work and maintenance happens.
+
+## What does success look like?
+
+- Stable libraries that are kept up to date with the core features
+- Good developer experience for the libraries
+- Reliable contractors that don't require too much management or excessive PR reviewing
+- Not spending too much money
+
+## Open questions
+
+- Which team should be responsible for reviewing the PRs for core features?
+- Would we want a contractor implementing autocapture for mobile?
+- We seem to have posthog-js-lite and posthog-js, what’s the plan and timeline for these? Do we want contractor support? (This seems more core)
+- For test coverage, do we want to aim for a % test coverage? Are we also wanting cypress style tests?
+- What profile of contractor do we want? Ideally we’d have some overlap of people e.g. 1 person to do all the mobile libraries, and up to 2 people for the server libraries
+    - I’d assume we’d want a senior engineer
+- How do we want to solve: “Cross compatibility nuances (kotlin -> java, swift -> objc)”? Do we want to build a swift library?
```

Comment:
I actually mostly built a swift library whilst working on Mobile session recordings.

Again, the challenge isn't the code here as much as it is coming up with a migration strategy / testing it in the field to get enough certainty that it is a stable rewrite.

Wrt Kotlin for example - this could make total sense as we approach Mobile Recordings and where an expert in the Android field would be great to advise and support on using all the best approaches